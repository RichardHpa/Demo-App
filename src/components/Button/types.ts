import type { variant } from 'types/variants'
import type { colors } from 'types/colors'

export type ButtonProps = {
  className?: string
  color?: colors
  variant?: variant
  disabled?: boolean
} & React.ComponentPropsWithoutRef<'button'>
