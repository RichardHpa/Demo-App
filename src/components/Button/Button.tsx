import clsx from 'clsx'
import React from 'react'

type ButtonProps = {
  className?: string
  color?: 'primary' | 'secondary' | 'danger'
  variant?: 'solid' | 'outline' | 'ghost'
  disabled?: boolean
} & React.ComponentPropsWithoutRef<'button'>

const classes = {
  base: 'inline-block py-2 px-4 rounded border',
  disabled: 'opacity-50 cursor-not-allowed pointer-events-none',
  variant: {
    solid: {
      primary: 'bg-blue-500 text-white hover:bg-blue-600 border-blue-500',
      secondary: 'bg-gray-500 text-white hover:bg-gray-600 border-gray-500',
      danger: 'bg-red-500 text-white hover:bg-red-600 border-red-500',
    },
    outline: {
      primary: 'text-blue-500 hover:bg-blue-500 hover:text-white border-blue-500',
      secondary: 'text-gray-500 hover:bg-gray-500 hover:text-white border-gray-500',
      danger: 'text-red-500 hover:bg-red-500 hover:text-white border-red-500',
    },
    ghost: {
      primary: 'text-blue-500 hover:bg-blue-500 hover:text-white border-transparent',
      secondary: 'text-gray-500 hover:bg-gray-500 hover:text-white border-transparent',
      danger: 'text-red-500 hover:bg-red-500 hover:text-white border-transparent',
    },
  },
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ color = 'primary', className, variant = 'solid', disabled, ...props }, ref) => {
    return (
      <button
        {...props}
        disabled={disabled}
        ref={ref}
        className={clsx(
          classes.base,
          classes.variant[variant][color],
          { [classes.disabled]: disabled },
          className,
        )}
      />
    )
  },
)

Button.displayName = 'Button'
