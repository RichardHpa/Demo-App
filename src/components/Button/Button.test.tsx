import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Button } from './Button'

describe('Button', () => {
  test('renders Button', () => {
    render(<Button>Test</Button>)
    const button = screen.getByRole('button', { name: /test/i })
    expect(button).toBeInTheDocument()

    expect(button).toMatchSnapshot()
  })

  test('renders Button with onClick', () => {
    const onClick = jest.fn()
    render(<Button onClick={onClick}>Test</Button>)
    userEvent.click(screen.getByRole('button'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  test('renders Button with disabled', () => {
    render(<Button disabled>Test</Button>)
    expect(screen.getByRole('button')).toBeDisabled()
  })

  test('renders Button with type', () => {
    render(<Button type='submit'>Test</Button>)
    expect(screen.getByRole('button')).toHaveAttribute('type', 'submit')
  })

  test.each([
    ['primary', 'bg-blue-500'],
    ['secondary', 'bg-gray-500'],
    ['danger', 'bg-red-500'],
  ] as const)('renders Button with color %s', (color, expected) => {
    render(<Button color={color}>Test</Button>)
    expect(screen.getByRole('button')).toHaveClass(expected)
  })

  test.each([
    ['solid', 'text-white'],
    ['outline', 'text-blue-500'],
    ['ghost', 'text-blue-500'],
  ] as const)('renders Button with variant %s', (variant, expected) => {
    render(<Button variant={variant}>Test</Button>)
    expect(screen.getByRole('button')).toHaveClass(expected)
  })
})
