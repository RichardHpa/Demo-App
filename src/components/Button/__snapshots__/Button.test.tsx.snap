// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders Button 1`] = `
<button
  class="inline-block py-2 px-4 rounded border bg-blue-500 text-white hover:bg-blue-600 border-blue-500"
>
  Test
</button>
`;
